lib/xapi-fdcaps/operations.ml

```diff
@@ -0,0 +1,206 @@
+(*
+ * Copyright (C) 2023 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Properties
+
+type +!'a props = {
+    props: ('b, 'c) Properties.props
+  ; custom_ftruncate: (int64 -> unit) option
+  ; fd: Safefd.t
+}
+  constraint 'a = ('b, 'c) Properties.props
+
+type +!'a t = 'a props constraint 'a = (_, _) Properties.t
+
+type (+!'a, +!'b) make = ('a, 'b) Properties.t t
+
+let dump ppf =
+  Fmt.(
+    Dump.(
+      record
+        [
+          field "props" (fun t -> t.props) pp
+        ; field "custom_ftruncate"
+            (fun t -> Option.is_some t.custom_ftruncate)
+            bool
+        ; field "fd" (fun t -> t.fd) Safefd.dump
+        ]
+    )
+  )
+    ppf
+
+let pp ppf =
+  Fmt.(
+    record
+      ~sep:Fmt.(any "; ")
+      [
+        field "props" (fun t -> t.props) pp
+      ; field "custom_ftruncate"
+          (fun t -> Option.is_some t.custom_ftruncate)
+          bool
+      ; field "fd" (fun t -> t.fd) Safefd.pp
+      ]
+  )
+    ppf
+
+let close t = Safefd.idempotent_close_exn t.fd
+
+let with_fd t f =
+  let finally () = close t in
+  Fun.protect ~finally (fun () -> f t)
+
+module Syntax = struct let ( let@ ) f x = f x end
+
+open Syntax
```

Comment:
```suggestion
let ( let@ ) f x = f x
```

Reply:
Syntax is in a separate module in the mli, you mean I should move the let declaration to the toplevel?
I thought to avoid clashes in the toplevel module (although it is unlikely that we'd define `let@` with a different semantics, clashes are more likely for the monadic let versions)